### PR TITLE
Rappel system

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request

**Please TM before merging. Numbers here are very subject to change.**
Revival of https://github.com/tgstation/TerraGov-Marine-Corps/pull/10386, with my own changes.

---

Adds a rappel system, attachable to the tadpole dropship equipment slot. Adds a spare replacement cord box.
'Rines get one of each to start. Replacement systems can be fabricated for 300 points each. Extra replacement cords can be bought at req for 30 points each.

- **Deploying the rappel.** When equipped, tadpole can deploy a rappel rope to any tile, as long as it is not underground or dense. Deploying takes 5 seconds and can be cancelled.

- **Retracting the rappel.** The rappel can then be retracted by pressing the designate icon again. This takes 5 seconds, and _is not cancelable_ to prevent cheese. The rappel cannot be retracted if it is hooked by a xeno attack. The rappel also automatically retracts if the tadpole lands or is sent back to ship.

- **Rappelling down.** Clicking on the rappel system in the tadpole when it's deployed rappels you down after 5 seconds. This is cancelable. It also provides visual and audio feedback for the poor xeno at the lower end so they don't get instakilled while resting.

- **Rappelling back up.** When a human clicks on a deployed rappel rope groundside, after a 5 second do_after, they are lifted up into the tadpole. When someone is lifted, the rope retracts with them - it must then be re-deployed manually. This _cooould_ lead to some situations where the tadpole is stranded in space with a dead 'rine in it, or where a xeno sneaks up and cuts the rappel with nobody in the tad, but then again, should the pilot really be leaving it unattended? 

Note that **any items or bodies dragged or fireman carried when rappelling up are not taken with you.** It's not possible to evac an entire dead push; you'll still need to land tadpole as-is for that. Similarly, if you rappel down and die, chances are tadpole can't save you. 
Good luck!

- **Disabling the rappel.** When a xeno clicks on a deployed rappel rope, after a 5 second do_after, they tug on the rope so hard it breaks the rappel system! This forces the rappel to retract. It also **disables** the system and **destroys the rappel cord**. This completely bricks the system until it is fully repaired.

- **Repairing the rappel.** The rappel system has two conditions for repair. Firstly, its mechanisms must be repaired - its self-repair functions take care of this after (currently) 10 minutes. You'll know when the mechanisms are damaged because it'll be smoking. Secondly, its cord must be replaced. Clicking on the system with a spare cord box will replace the cord, but this will use up the box in the process. You'll know when a cord needs replacing because a red exclamation mark will be visible on the system.

Official lore diagram for disabling the rappel, so you know it's fully realistic:
image
## Why It's Good For The Game

Gives marines a tool to execute more advanced strategies as a team, _but which the tadpole has to commit to._ No MG, no operating table, just pure infantry goodness, as the founding coders intended. Gives more things for PO to do and keeps xenos on their toes while (hopefully) giving them some proper counterplay to it. Lets tadpole finally be able to drop in FOB without crushing everything.

The code for this was made before the warlock PR was even made, but now that it's out, this could provide a good way for tadpole to assist QRF and flanks without getting immediately fucked by 3 AoEs. Warlock is still useful against it, though - it's a bit hard to climb back up to safety and avoid going perma when you're being thrown more than a cave push. 

This makes it risky, but filling a niche nothing else currently does. As Tye's PR says, this will really expand tadpole playstyles and I can't wait to see what creative plays are possible with it.
## Changelog
:cl: tyeagg for original code, breadhunt for reviving it
add: Adds rappel system, turning POs into glorified elevators once more.
add: Adds the spare rappel cord box. Required to repair the rappel system.
imageadd: rappel system sprite, rappel cord box
/:cl:

### CODE REVIEW

**Comment by Losenis** [comment]
Rather than a full replacement you could have the rappel system use up a "rope" item that is spent on breakage and can be printed, an item small enough to be carried on hand rather than a ripley. That way you can control the cost of the system itself and the spare rope, since you also have the possibilities of ~~marines with grenades~~ xenos breaking the system itself rather than just the rope. It'd enable carrying a finite amount of replacements on the tadpole that'd necessitate someone on it to use.

**Comment by MereWalker** [comment]
This is certainly interesting.... look forward to it.

File: code/game/objects/structures/dropship_equipment.dm [hunk 1]
```diff
@@ -1030,3 +1033,289 @@
...
+	origin.deploy_rope(target_turf)
+	playsound(origin, 'sound/machines/chime.ogg', 50, FALSE)
+
+
+
+//The actual system you put on the tadpole
+/obj/structure/dropship_equipment/rappel_system
+	equip_category = DROPSHIP_CREW_WEAPON
+	name = "rappel deployment system"
+	desc = "A system that deploys rappel ropes to go up or down fast, without the need for the tadpole to land. You need to designate the rappel point at the navigation computer."
```
Comment (XSlayer300) [inline code/game/objects/structures/dropship_equipment.dm:1081]:
```suggestion
	desc = "A system that deploys rappel ropes to go up or down fast, without the need for the Tadpole to land. You need to designate the rappel point at the navigation computer."
```

**Comment by bluezorua** [comment]
I don't think the rapel should perma break, or at least have req buyable replacement ropes or something 

a heavily teamwork focused item shouldn't be lost if a single person is stupid

**Comment by breadhunt** [comment]
My computer fucking broke, going to reopen this when I get a new one

Extra ropes sound like a cool idea and a good compromise though, I'd be interested in coding it up when I fucking can

The coder curse strikes again

**Comment by breadhunt** [comment]
Managed to jerryrig my laptop into existence once more, lets go

Rappel rope as a replacement is a very cool idea. 30 req points + the current 5 minute timer sounds like a decent cost, thoughts?

You're completely right about "entire team shouldn't be punished" @bluezorua, but at the same time I think it needs to be punish*able* by zenos. The PO is still part of the team, and defending the rope is also the responsibility of the deploying marines.

With this, destroying the rope leaves deployed rines vulnerable and provides a cost for successive mistakes with rappel placement while not completely locking the PO out of using the rappel for the entire round. Though that last bit might not be a bad idea still - rappel is hard to punish by nature of it being a 4 pixel wide rope.

**Review by Kuro020 — APPROVED** [review]

**Comment by S1tr3p** [comment]
THIS IS FUCKING AWESOME!!!!!

File: code/game/objects/structures/dropship_equipment.dm [hunk 2]
```diff
@@ -1039,3 +1042,328 @@
...
+/datum/action/innate/rappel_designate/Activate()
+	var/mob/C = target
+	var/mob/camera/aiEye/remote/remote_eye = C.remote_control
+	var/turf/target_turf = get_turf(remote_eye)
+	var/area/target_area = get_area(target_turf)
+
+	if(origin.retracting) //We're already retracting a rope!
+		return
+
+	if(origin.disabled || origin.cord_damaged || target_turf.density || target_area.ceiling >= CEILING_DEEP_UNDERGROUND)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1064]: Shouldn't this just be CEILING_UNDERGROUND?

File: code/game/objects/structures/dropship_equipment.dm [hunk 3]
```diff
@@ -1039,3 +1042,328 @@
...
+	var/obj/structure/dropship_equipment/rappel_system/origin
+
+/datum/action/innate/rappel_designate/Activate()
+	var/mob/C = target
+	var/mob/camera/aiEye/remote/remote_eye = C.remote_control
+	var/turf/target_turf = get_turf(remote_eye)
+	var/area/target_area = get_area(target_turf)
+
+	if(origin.retracting) //We're already retracting a rope!
+		return
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1062]: You might as well put the early returns as early as possible, i.e. before the vars where possible.

File: code/game/objects/structures/dropship_equipment.dm [hunk 4]
```diff
@@ -1039,3 +1042,328 @@
...
+
+	if(origin.disabled || origin.cord_damaged || target_turf.density || target_area.ceiling >= CEILING_DEEP_UNDERGROUND)
+		playsound(origin, 'sound/machines/buzz-two.ogg', 25)
+		return
+	if(origin.hooked)
+		playsound(origin, 'sound/machines/buzz-sigh.ogg', 25)
+		return
+
+	C.playsound_local(origin, 'sound/effects/binoctarget.ogg', 35)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1071]: Should all these feedback sounds also have a feedback message/balloon text to actual say whats going on?

File: code/game/objects/structures/dropship_equipment.dm [hunk 5]
```diff
@@ -1039,3 +1042,328 @@
...
+	///Whether xenos have temporarily disabled the system or not
+	var/disabled = FALSE
+	///Whether the system's cord needs replacing or not
+	var/cord_damaged = FALSE
+	///Smoke particle holder for when it's disabled
+	var/obj/effect/abstract/particle_holder/disabled_smoke
+	///Whether the system is currently retracting a rope or not
+	var/retracting = FALSE
+	///Whether a xeno is currently disabling the system or not
+	var/hooked = FALSE
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1111]: Do all these binary vars actually need to exist? They could probs be just flags, or potentially just a general state var.

File: code/game/objects/structures/dropship_equipment.dm [hunk 6]
```diff
@@ -1039,3 +1042,328 @@
...
+			return
+		balloon_alert(user, "You start replacing the rappel cord...")
+		if(!do_after(user, 2 SECONDS, TRUE, src, BUSY_ICON_GENERIC))
+			return
+		cord_damaged = FALSE
+		icon_state = "rappel_hatch_locked"
+		balloon_alert(user, "You replace the rappel cord.")
+		QDEL_NULL(I)
+
+	//TODO: make this call attack_hand if they're human - and for ghosts
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1202]: todone?

File: code/game/objects/structures/dropship_equipment.dm [hunk 7]
```diff
@@ -1039,3 +1042,328 @@
...
+/obj/structure/dropship_equipment/rappel_system/examine(mob/user)
+	. = ..()
+	if(disabled)
+		. += "It's smoking. It'll take some time for its self-repair function to kick in."
+	if(cord_damaged)
+		. += "The rappel's cord is completely broken and needs replacing."
+
+
+///Human interaction with the rappel system; this is how people rappel down
+/obj/structure/dropship_equipment/rappel_system/attack_hand(mob/living/carbon/human/user)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1134]: you've typecast user as human when its normally just living, but is anything actually stopping nonhuman users from using this? Also I'm pretty sure this should call parent.

File: code/game/objects/structures/dropship_equipment.dm [hunk 8]
```diff
@@ -1039,3 +1042,328 @@
...
+			return
+		cord_damaged = FALSE
+		icon_state = "rappel_hatch_locked"
+		balloon_alert(user, "You replace the rappel cord.")
+		QDEL_NULL(I)
+
+	//TODO: make this call attack_hand if they're human - and for ghosts
+
+///Human animation for dropping down
+/obj/structure/dropship_equipment/rappel_system/proc/rappel_animation(mob/living/carbon/human/user)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1205]: ditto

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/dropship_equipment.dm (hunk 5 above)
Comment (BraveMole) [reply code/game/objects/structures/dropship_equipment.dm:1111]: Def should be state var / flags

File: code/game/objects/structures/dropship_equipment.dm [hunk 9]
```diff
@@ -1039,3 +1042,328 @@
...
+	///Smoke particle holder for when it's disabled
+	var/obj/effect/abstract/particle_holder/disabled_smoke
+	///Whether the system is currently retracting a rope or not
+	var/retracting = FALSE
+	///Whether a xeno is currently disabling the system or not
+	var/hooked = FALSE
+
+/obj/structure/dropship_equipment/rappel_system/Initialize()
+	. = ..()
+	rope = new
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1115]: Rope = new(src), and the rope parent system should be set in rope constructor

File: code/game/objects/structures/dropship_equipment.dm [hunk 10]
```diff
@@ -1039,3 +1042,328 @@
...
+	if(target_turf.density)
+		balloon_alert(user, "You can't rappel into a wall!")
+		return
+
+	var/area/target_area = get_area(target_turf)
+	if(target_area.ceiling >= CEILING_DEEP_UNDERGROUND)
+		balloon_alert(user, "The rappel is too deep underground!")
+		return
+
+	rope.icon_state = "rope_rappeling"
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1165]: No icon state change outside of update_icon_state()

Also tgmc is oop design, stop accessing directly in rope vars where that should be function call

File: code/game/objects/structures/dropship_equipment.dm [hunk 11]
```diff
@@ -1039,3 +1042,328 @@
...
+	if(icon_state != "rappel_hatch_open")
+		flick("rappel_hatch_opening", src)
+		icon_state = "rappel_hatch_open"
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1168]: Ditto. And anyway you are flicking it and setting it, it doesnt make sense

File: code/game/objects/structures/dropship_equipment.dm [hunk 12]
```diff
@@ -1039,3 +1042,328 @@
...
+	if(do_after(user, 5 SECONDS, TRUE, target_turf, BUSY_ICON_GENERIC) && !user.lying_angle && !user.anchored && unlocked && !disabled && !cord_damaged)
+		playsound(target_turf, 'sound/effects/rappel.ogg', 50, TRUE)
+		playsound(src, 'sound/effects/rappel.ogg', 50, TRUE)
+		user.forceMove(target_turf)
+		rappel_animation(user)
+
+	flick("rappel_hatch_closing", src)
+	icon_state = "rappel_hatch_unlocked"
+	rope.icon_state = "rope"
+
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1183]: Ditto

File: code/game/objects/structures/dropship_equipment.dm [hunk 13]
```diff
@@ -1039,3 +1042,328 @@
...
+		QDEL_NULL(I)
+
+	//TODO: make this call attack_hand if they're human - and for ghosts
+
+///Human animation for dropping down
+/obj/structure/dropship_equipment/rappel_system/proc/rappel_animation(mob/living/carbon/human/user)
+	var/pre_rappel_alpha = user.alpha
+	user.alpha = 20
+	user.dir = WEST
+	user.canmove = FALSE
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1209]: Never use can move. Use traits or effects

File: code/game/objects/structures/dropship_equipment.dm [hunk 14]
```diff
@@ -1039,3 +1042,328 @@
...
+	name = "Designate rappel point"
+	action_icon = 'icons/mecha/actions_mecha.dmi'
+	action_icon_state = "mech_zoom_on"
+	var/obj/structure/dropship_equipment/rappel_system/origin
+
+/datum/action/innate/rappel_designate/Activate()
+	var/mob/C = target
+	var/mob/camera/aiEye/remote/remote_eye = C.remote_control
+	var/turf/target_turf = get_turf(remote_eye)
+	var/area/target_area = get_area(target_turf)
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1059]: Why are you setting these vars if there is a possibility you do not use them.
Define vars latest as possible

File: code/game/objects/structures/dropship_equipment.dm [hunk 15]
```diff
@@ -1039,3 +1042,328 @@
...
 	icon_state = "table2-idle"
+
+//////////////// RAPPEL SYSTEM /////////////////
+
+//Rappel target selection action
+/datum/action/innate/rappel_designate
+	name = "Designate rappel point"
+	action_icon = 'icons/mecha/actions_mecha.dmi'
+	action_icon_state = "mech_zoom_on"
+	var/obj/structure/dropship_equipment/rappel_system/origin
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1053]: Autodoc

File: code/game/objects/structures/dropship_equipment.dm [hunk 16]
```diff
@@ -1039,3 +1042,328 @@
...
+
+/datum/action/innate/rappel_designate/Activate()
+	var/mob/C = target
+	var/mob/camera/aiEye/remote/remote_eye = C.remote_control
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1057]: Whzt is the point of that var ?

File: code/game/objects/structures/dropship_equipment.dm [hunk 17]
```diff
@@ -1039,3 +1042,328 @@
...
+///This is the rope that the system deploys, a subtype of the HvH deployment rappel.
+///Created by the rappel system on init and stored in the rappel system when it's not in use
+/obj/effect/rappel_rope/tadpole
+	icon = 'icons/Marine/mainship_props.dmi'
+	name = "tadpole rappel rope"
+	light_system = STATIC_LIGHT
+	light_power = 0.5
+	light_range = 2
+	///The rappel system this rope originates from
+	var/obj/structure/dropship_equipment/rappel_system/parent_system
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1284]: You need to clean you vars on destructor

File: code/game/objects/structures/dropship_equipment.dm [hunk 18]
```diff
@@ -1039,3 +1042,328 @@
...
+	desc = "A system that deploys rappel ropes to go up or down fast, without the need for the Tadpole to land. You need to designate the rappel point at the navigation computer."
+	dropship_equipment_flags = IS_INTERACTABLE
+	icon_state = "rappel_hatch_locked"
+	point_cost = 300
+	coverage = 0
+	max_integrity = 300
+	anchored = TRUE
+	density = FALSE
+	///The rappeling rope we use
+	var/obj/effect/rappel_rope/tadpole/rope
```
Comment (BraveMole) [inline code/game/objects/structures/dropship_equipment.dm:1099]: Var cleaning is missing

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by RaidarV** [comment]
thabks vreadhunt berry cool

File: code/game/objects/structures/dropship_equipment.dm (hunk 4 above)
Comment (breadhunt) [reply code/game/objects/structures/dropship_equipment.dm:1071]: These are for the person in the camera actually designating the spot, who can't see balloon alerts on the system & won't be able to reliably see balloon alerts on NVG. I get your point, though - would making the feedback sounds local to the pilot be enough?

**Pull request closed without merging**